From b7b59910081cbc71834154e7659b7a37894a9dab Mon Sep 17 00:00:00 2001
From: Mahesh Vaidya <mahesh.vaidya@altera.com>
Date: Wed, 29 Apr 2026 13:32:09 -0700
Subject: [PATCH] Add PCIe RP driver support for Legacy INTx interrupts

Add PCIe RP driver support for Legacy INTx interrupts in Agilex.

Upstream-Status: Pending
Signed-off-by: Mahesh Vaidya <mahesh.vaidya@altera.com>
---
 drivers/pci/controller/pcie-altera.c | 42 ++++++++++++++++++++++------
 1 file changed, 34 insertions(+), 8 deletions(-)

diff --git a/drivers/pci/controller/pcie-altera.c b/drivers/pci/controller/pcie-altera.c
index 3dbb7adc421c..1768900dc044 100644
--- a/drivers/pci/controller/pcie-altera.c
+++ b/drivers/pci/controller/pcie-altera.c
@@ -84,7 +84,9 @@
 
 #define AGLX_BDF_REG			0x00002004
 #define AGLX_ROOT_PORT_IRQ_STATUS	0x14c
+#define AGLX_INTX_STATUS_ALL		0xf
 #define AGLX_ROOT_PORT_IRQ_ENABLE	0x150
+#define AGLX_INTX_ENA_ALL		0xf
 #define CFG_AER				BIT(4)
 
 #define AGLX_CFG_TARGET			GENMASK(13, 12)
@@ -828,7 +830,8 @@ static void aglx_isr(struct irq_desc *desc)
 	struct irq_chip *chip = irq_desc_get_chip(desc);
 	struct altera_pcie *pcie;
 	struct device *dev;
-	u32 status;
+	unsigned long status;
+	u32 bit;
 	int ret;
 
 	chained_irq_enter(chip, desc);
@@ -838,13 +841,23 @@ static void aglx_isr(struct irq_desc *desc)
 	status = readl(pcie->hip_base + pcie->pcie_data->port_conf_offset +
 		       pcie->pcie_data->port_irq_status_offset);
 
-	if (status & CFG_AER) {
-		writel(CFG_AER, (pcie->hip_base + pcie->pcie_data->port_conf_offset +
-				 pcie->pcie_data->port_irq_status_offset));
+	if ((status & AGLX_INTX_STATUS_ALL) != 0) {
+		for_each_set_bit(bit, &status, PCI_NUM_INTX) {
+			/* clear INTx interrupt */
+			writel(1<<bit, pcie->hip_base + pcie->pcie_data->port_conf_offset +
+			pcie->pcie_data->port_irq_status_offset);
+			ret = generic_handle_domain_irq(pcie->irq_domain, bit);
+			if (ret)
+				dev_err_ratelimited(dev, "unexpected IRQ, INT%d\n", bit);
+		}
+	}
 
+	if (status & CFG_AER) {
 		ret = generic_handle_domain_irq(pcie->irq_domain, 0);
 		if (ret)
 			dev_err_ratelimited(dev, "unexpected IRQ %d\n", pcie->irq);
+		writel(CFG_AER, (pcie->hip_base + pcie->pcie_data->port_conf_offset +
+				 pcie->pcie_data->port_irq_status_offset));
 	}
 	chained_irq_exit(chip, desc);
 }
@@ -852,12 +865,24 @@ static void aglx_isr(struct irq_desc *desc)
 static int altera_pcie_init_irq_domain(struct altera_pcie *pcie)
 {
 	struct device *dev = &pcie->pdev->dev;
+	struct device_node *node = dev->of_node;
+	struct device_node *legacy_intc_np;
 
 	/* Setup INTx */
-	pcie->irq_domain = irq_domain_create_linear(dev_fwnode(dev), PCI_NUM_INTX,
-					&intx_domain_ops, pcie);
+	legacy_intc_np = of_get_child_by_name(node, "legacy-interrupt-controller");
+	if (!legacy_intc_np) {
+		dev_err(dev, "Failed to find legacy interrupt controller node\n");
+		return -ENODEV;
+	}
+
+	pcie->irq_domain = irq_domain_create_linear(of_fwnode_handle(legacy_intc_np),
+						    PCI_NUM_INTX,
+						    &intx_domain_ops, pcie);
+
+	of_node_put(legacy_intc_np);
+
 	if (!pcie->irq_domain) {
-		dev_err(dev, "Failed to get a INTx IRQ domain\n");
+		dev_err(dev, "Failed to get INTx IRQ domain\n");
 		return -ENOMEM;
 	}
 
@@ -1028,7 +1053,8 @@ static int altera_pcie_probe(struct platform_device *pdev)
 		cra_writel(pcie, P2A_INT_ENA_ALL, P2A_INT_ENABLE);
 		altera_pcie_host_init(pcie);
 	} else if (pcie->pcie_data->version == ALTERA_PCIE_V3) {
-		writel(CFG_AER,
+		/* enable all interrupts */
+		writel((CFG_AER | AGLX_INTX_ENA_ALL),
 		       pcie->hip_base + pcie->pcie_data->port_conf_offset +
 		       pcie->pcie_data->port_irq_enable_offset);
 	}
-- 
2.34.1

